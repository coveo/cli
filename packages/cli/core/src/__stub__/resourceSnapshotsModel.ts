import {
  ResourceSnapshotsModel,
  ResourceSnapshotsReportModel,
} from '@coveo/platform-client';

export const getDummySnapshotModel = (
  orgId: string,
  snapshotId: string,
  reports: ResourceSnapshotsReportModel[] = []
): ResourceSnapshotsModel => ({
  id: snapshotId,
  createdBy: 'user@coveo.com',
  createdDate: 1622555047116,
  targetId: orgId,
  developerNote: 'hello',
  reports: reports,
  synchronizationReports: [],
  contentSummary: {EXTENSION: 1, FIELD: 11},
});
