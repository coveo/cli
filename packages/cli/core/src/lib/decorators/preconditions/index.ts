// TODO: Remove, unusued
export * from './git';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './node';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './npx';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './npm';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './ng';
