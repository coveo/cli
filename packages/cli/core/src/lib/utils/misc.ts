import {resolve} from 'path';
import {valid} from 'semver';

export function getPackageVersion(packageName: string) {
  const pathToPackageJson = resolve(
    __dirname,
    ...new Array(3).fill('..'),
    'package.json'
  );
  const pkg = require(pathToPackageJson);
  const dep =
    pkg.dependencies[packageName] ||
    pkg.devDependencies[packageName] ||
    pkg.peerDependencies[packageName];
  const defaultVersion = valid(dep?.toString())!;

  return defaultVersion;
}
