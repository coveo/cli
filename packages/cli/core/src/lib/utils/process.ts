import {spawn, SpawnOptions, ChildProcess} from 'child_process';

function isErrnoException(error: Error): error is NodeJS.ErrnoException {
  return Object.hasOwnProperty.call(error, 'errno');
}

/**
 *
 * @private
 * @param {string} command The command to run
 * @param {string[]} args List of string arguments
 * @param {SpawnOptions} [options={}]
 * @returns {Promise<number>} A promise of the the child process exit code
 */
export function spawnProcess(
  command: string,
  args: string[],
  options: SpawnOptions = {}
): Promise<number> {
  return new Promise((resolve, reject) => {
    spawn(command, args, {
      stdio: ['inherit', 'inherit', 'inherit'],
      ...options,
    }).on('close', (code) => {
      if (code !== 0) {
        // TODO: inject logger decorator so this util method can log
        // this.debug(
        //   `Following command has failed: ${command} ${args.join(' ')}`
        // );

        // Any error message triggered by the execution of the subprocess will be
        // displayed since the terminal is shared by both parent and child processes.
        // So no need to bubble up the error.
        reject(code);
      } else {
        resolve(0);
      }
    });
  });
}

export interface SpawnProcessOutput {
  stdout: string;
  stderr: string;
  exitCode?: string;
}

export async function spawnProcessOutput(
  command: string,
  args: string[],
  options: SpawnOptions = {}
): Promise<SpawnProcessOutput> {
  return new Promise((resolve) => {
    const output: SpawnProcessOutput = {
      stdout: '',
      stderr: '',
    };

    const child = spawn(command, args, options);

    child.stdout?.on('data', (d) => {
      output.stdout += d;
    });

    child.stderr?.on('data', (d) => {
      output.stderr += d;
    });

    const closeListener = (code: number | null) => {
      output.exitCode = Number.isInteger(code) ? String(code) : undefined;
      resolve(output);
    };

    child.on('close', closeListener);

    child.on('error', (e) => {
      output.stderr += e.message;
      if (isErrnoException(e) && e.code) {
        child.off('close', closeListener);
        output.exitCode = e.code;
        resolve(output);
      }
    });
  });
}

export const handleForkedProcess = (subprocess: ChildProcess) => {
  return new Promise<void>((resolve, reject) => {
    subprocess.on('exit', () => {
      resolve();
    });
    subprocess.on('message', (message) => {
      reject(message);
    });
  });
};
