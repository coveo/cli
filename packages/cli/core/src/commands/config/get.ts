import {CLICommand} from '@coveo/cli-commons/command/cliCommand';
import {Config} from '@coveo/cli-commons/config/config';
import {ConfigRenderer} from '@coveo/cli-commons/config/configRenderer';
import {Trackable} from '@coveo/cli-commons/preconditions/trackable';
import type {Example} from '@oclif/core/lib/interfaces';

export default class Get extends CLICommand {
  public static description = 'Display the current Coveo CLI configuration.';

  public static args = [
    {
      name: 'key',
      description: 'The config key for which to show the value',
      required: false,
    },
  ];

  public static examples: Example[] = [
    {
      command: 'coveo config:get',
      description: 'Get all the configuration values',
    },
    {
      command: 'coveo config:get organization',
      description: 'Get the organization to which you are connected',
    },
    {
      command: 'coveo config:get accessToken',
      description: 'Get the access token given to you by the Coveo Platform',
    },
  ];

  @Trackable()
  public async run() {
    const {args} = await this.parse(Get);
    const cfg = new Config(this.config.configDir);
    const keysToRender = args.key ? [args.key] : undefined;
    ConfigRenderer.render(cfg, keysToRender);
  }
}
