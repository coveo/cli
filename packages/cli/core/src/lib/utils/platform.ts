import {Config} from '@coveo/cli-commons/config/config';

export function getTargetOrg(config: Config, target?: string) {
  if (target) {
    return target;
  }
  const cfg = config.get();
  return cfg.organization;
}
