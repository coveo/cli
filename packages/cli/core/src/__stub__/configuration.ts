import {Region} from '@coveo/platform-client';
import type {Config, Configuration} from '@coveo/cli-commons/config/config';
import {PlatformEnvironment} from '@coveo/cli-commons/platform/environment';

export const defaultConfiguration = {
  environment: PlatformEnvironment.Dev,
  organization: 'my-org',
  region: Region.US,
  accessToken: 'some-api-key-or-token',
  anonymous: false,
} as Configuration;

export const configurationMock: (
  configuration?: Configuration
) => () => Config =
  (configuration = defaultConfiguration) =>
  () =>
    ({
      get: () => configuration,
    } as Config);
