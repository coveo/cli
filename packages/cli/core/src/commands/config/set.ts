import {CLICommand} from '@coveo/cli-commons/command/cliCommand';
import {CliUx, Flags} from '@oclif/core';
import {Config} from '@coveo/cli-commons/config/config';
import {AuthenticatedClient} from '@coveo/cli-commons/platform/authenticatedClient';
import {
  IsAuthenticated,
  Preconditions,
} from '@coveo/cli-commons/preconditions/index';
import {Trackable} from '@coveo/cli-commons/preconditions/trackable';
import {InvalidCommandError} from '../../lib/errors/InvalidCommandError';
import {ConfigRenderer} from '@coveo/cli-commons/config/configRenderer';
import type {Example} from '@oclif/core/lib/interfaces';

export default class Set extends CLICommand {
  public static description = 'Modify the current Coveo CLI configuration.';

  public static flags = {
    // TODO CDX-1246
    environment: Flags.string({
      char: 'e',
      hidden: true,
    }),
    // TODO CDX-1246
    region: Flags.string({
      char: 'r',
      hidden: true,
    }),
    organization: Flags.string({
      char: 'o',
      description:
        'The identifier of the organization inside which to perform operations. See <https://docs.coveo.com/en/1562/#organization-id-and-other-information>.',
      helpValue: 'myOrgID',
    }),
  };

  public static examples: Example[] = [
    {
      command: 'coveo config:set --organization myOrgId',
      description: 'connect to a different organization',
    },
  ];

  @Trackable()
  @Preconditions(IsAuthenticated())
  public async run() {
    const {flags} = await this.parse(Set);
    // TODO CDX-1246
    if (flags.environment || flags.region) {
      CliUx.ux.error(
        'To connect to a different region or environment, use the `auth:login` command'
      );
    }
    if (Object.entries(flags).length === 0) {
      throw new InvalidCommandError('Command should contain at least 1 flag');
    }
    const cfg = new Config(this.config.configDir);
    if (flags.organization) {
      await this.verifyOrganization(flags.organization);
      cfg.set('organization', flags.organization);
    }

    ConfigRenderer.render(cfg);
  }

  private async verifyOrganization(org: string) {
    const hasAccess = await new AuthenticatedClient().getUserHasAccessToOrg(
      org
    );
    if (!hasAccess) {
      this.error(
        `You either don't have access to organization ${org}, or it doesn't exist.`
      );
    }
  }
}
