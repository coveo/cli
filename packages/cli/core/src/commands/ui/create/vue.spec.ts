jest.mock('../../../lib/decorators/preconditions/npx');
jest.mock('../../../lib/decorators/preconditions/node');
jest.mock('@coveo/cli-commons/preconditions/apiKeyPrivilege');
jest.mock('../../../lib/utils/process');
jest.mock('../../../lib/oauth/oauth');
jest.mock('@coveo/cli-commons/config/config');
jest.mock('@coveo/cli-commons/preconditions/trackable');

jest.mock('@coveo/cli-commons/platform/authenticatedClient');
jest.mock('../../../lib/utils/misc');
jest.mock('@coveo/platform-client');

import {test} from '@oclif/test';
import {spawnProcess} from '../../../lib/utils/process';
import {AuthenticatedClient} from '@coveo/cli-commons/platform/authenticatedClient';
import PlatformClient from '@coveo/platform-client';
import {Config} from '@coveo/cli-commons/config/config';
import {
  IsNodeVersionInRange,
  IsNpxInstalled,
} from '../../../lib/decorators/preconditions/index';
import {HasNecessaryCoveoPrivileges} from '@coveo/cli-commons/preconditions/index';
import {getPackageVersion} from '../../../lib/utils/misc';
import {configurationMock} from '../../../__stub__/configuration';
import {mockPreconditions} from '@coveo/cli-commons/preconditions/mockPreconditions';

describe('ui:create:vue', () => {
  const mockedConfig = jest.mocked(Config);
  const mockedSpawnProcess = jest.mocked(spawnProcess);
  const mockedPlatformClient = jest.mocked(PlatformClient);
  const mockedGetPackageVersion = jest.mocked(getPackageVersion);
  const mockedAuthenticatedClient = jest.mocked(AuthenticatedClient);
  const mockedIsNpxInstalled = jest.mocked(IsNpxInstalled);
  const mockedIsNodeVersionInRange = jest.mocked(IsNodeVersionInRange);
  const vueCliPackage = '@vue/cli';
  const mockedApiKeyPrivilege = jest.mocked(HasNecessaryCoveoPrivileges);
  const mockedCreateImpersonateApiKey = jest.fn();
  const preconditionStatus = {
    node: true,
    npx: true,
    apiKey: true,
  };
  const doMockPreconditions = function () {
    const mockedPreconditions = mockPreconditions(preconditionStatus);
    mockedIsNodeVersionInRange.mockReturnValue(mockedPreconditions.node);
    mockedIsNpxInstalled.mockReturnValue(mockedPreconditions.npx);
    mockedApiKeyPrivilege.mockReturnValue(mockedPreconditions.apiKey);
  };

  const doMockSpawnProcess = () => {
    mockedSpawnProcess.mockResolvedValue(0);
  };

  const doMockedGetPackageVersion = () => {
    mockedGetPackageVersion.mockReturnValue('1.0.0');
  };

  const doMockConfiguration = () => {
    mockedConfig.mockImplementation(configurationMock());
  };

  const doMockAuthenticatedClient = () => {
    mockedCreateImpersonateApiKey.mockImplementation((_name: string) =>
      Promise.resolve({value: 'foo'})
    );

    mockedAuthenticatedClient.mockImplementation(
      () =>
        ({
          createImpersonateApiKey: mockedCreateImpersonateApiKey,
          getUsername: () => Promise.resolve('bob@coveo.com'),
          getClient: () =>
            Promise.resolve(
              mockedPlatformClient.getMockImplementation()!({
                accessToken: 'foo',
                organizationId: 'my-org',
              })
            ),
          cfg: mockedConfig.getMockImplementation()!('./'),
        } as unknown as AuthenticatedClient)
    );
  };

  const doMockPlatformClient = () => {
    mockedPlatformClient.mockImplementation(
      () =>
        ({
          initialize: () => Promise.resolve(),
        } as PlatformClient)
    );
  };

  beforeEach(() => {
    doMockedGetPackageVersion();
    doMockSpawnProcess();
    doMockPlatformClient();
    doMockConfiguration();
    doMockAuthenticatedClient();
    doMockPreconditions();
    preconditionStatus.node = true;
    preconditionStatus.npx = true;
    preconditionStatus.apiKey = true;
  });

  afterEach(() => {
    mockedIsNodeVersionInRange.mockClear();
    mockedApiKeyPrivilege.mockClear();
  });

  test
    .stdout()
    .stderr()
    .do(() => {
      preconditionStatus.apiKey = false;
    })
    .command(['ui:create:vue', 'myapp'])
    .catch(/apiKey Precondition Error/)
    .it(
      'should not execute the command if the API key preconditions are not respected'
    );

  test
    .stdout()
    .stderr()
    .do(() => {
      preconditionStatus.node = false;
    })
    .command(['ui:create:vue', 'myapp'])
    .catch(/node Precondition Error/)
    .it(
      'should not execute the command if the preconditions are not respected'
    );

  test
    .stdout()
    .stderr()
    .command(['ui:create:vue'])
    .catch(/Missing 1 required arg/)
    .it('requires application name argument');

  test
    .stdout()
    .stderr()
    .command(['ui:create:vue', 'myapp'])
    .it('should start 2 spawn processes with default preset arguments', () => {
      expect(mockedSpawnProcess).toHaveBeenCalledTimes(2);
      expect(mockedSpawnProcess).nthCalledWith(
        1,
        expect.stringContaining('npx'),
        [
          `${vueCliPackage}@1.0.0`,
          'create',
          'myapp',
          '--inlinePreset',
          expect.objectContaining({}),
          '--skipGetStarted',
          '--bare',
        ],
        expect.objectContaining({})
      );
      expect(mockedSpawnProcess).nthCalledWith(
        2,
        expect.stringContaining('npx'),
        [
          `${vueCliPackage}@1.0.0`,
          'add',
          '@coveo/vue-cli-plugin-typescript@1.0.0',
          '--orgId',
          'my-org',
          '--apiKey',
          'foo',
          '--platformUrl',
          'https://platformdev.cloud.coveo.com',
          '--user',
          'bob@coveo.com',
        ],
        expect.objectContaining({cwd: 'myapp'})
      );
    });
});
