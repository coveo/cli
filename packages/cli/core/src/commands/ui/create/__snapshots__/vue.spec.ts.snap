// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui:create:vue when a directory with the projectName exists when the directory is empty should write the .env file 1`] = `
[
  "/foo/bar/myapp/.env",
  "VITE_COVEO_PLATFORM_URL=https://platformdev.cloud.coveo.com
VITE_COVEO_ORGANIZATION_ID=my-org
VITE_COVEO_USER_EMAIL=bob@coveo.com
SERVER_COVEO_API_KEY=foo",
]
`;

exports[`ui:create:vue when a directory with the projectName exists when the directory is not empty should exit with an error 1`] = `"/foo/bar/myapp is a non-empty directory, use an empty directory or unused name for your project"`;

exports[`ui:create:vue when a file with the projectName exists should exit with an error 1`] = `"/foo/bar/myapp already exists and is not a directory, use an unused directory name for your project"`;

exports[`ui:create:vue when the projectName does not exists yet should write the .env file 1`] = `
[
  "/foo/bar/myapp/.env",
  "VITE_COVEO_PLATFORM_URL=https://platformdev.cloud.coveo.com
VITE_COVEO_ORGANIZATION_ID=my-org
VITE_COVEO_USER_EMAIL=bob@coveo.com
SERVER_COVEO_API_KEY=foo",
]
`;
